backend/src/main/java/com/google/sps/TripCRUD.java

```
@@ -0,0 +1,138 @@
+    Key entityKey = KeyFactory.createKey("Trip", Long.parseLong(tripId));
+    return datastore.get(entityKey);
+  }
+
+  /**
+   * Converts Trip Entity to Json
+   *
+   * @param tripEntity Trip Entity
+   * @return JsonObject of trip
+   */
+  public static JsonObject toJson(Entity tripEntity) {
+    JsonObject jsonTrip = new JsonObject();
+    jsonTrip.addProperty("tripId", Long.toString(tripEntity.getKey().getId()));
+    jsonTrip.addProperty(
+        "isOptimized", Boolean.parseBoolean(tripEntity.getProperty("isOptimized").toString()));
+    jsonTrip.addProperty("searchText", tripEntity.getProperty("searchText").toString());
+    jsonTrip.addProperty("tripName", tripEntity.getProperty("tripName").toString());
+    ArrayList<JsonObject> attractions = new ArrayList<JsonObject>();
+    for (EmbeddedEntity attraction :
+        (ArrayList<EmbeddedEntity>) tripEntity.getProperty("attractions")) {
+      JsonObject attractionJson = new JsonObject();
+      attractionJson.addProperty(
+          "attractionName", attraction.getProperty("attractionName").toString());
+      attractionJson.addProperty(
+          "photoReference", attraction.getProperty("photoReference").toString());
+      attractionJson.addProperty(
+          "routeIndex", Integer.parseInt(attraction.getProperty("routeIndex").toString()));
+      attractionJson.addProperty("coordinates", attraction.getProperty("coordinates").toString());
+      attractions.add(attractionJson);
+    }
+    jsonTrip.addProperty("attractions", attractions.toString());
+    return jsonTrip;
+  }
+
+  /**
+   * Updates a trip's properties
+   *
+   * @param tripId id of the trip to find
+   * @param tripData string representation of tripData json
+   */
+  public static void updateTrip(String tripId, String tripData) {
+    Entity tripEntity;
+    try {
+      tripEntity = TripCRUD.readTrip(tripId);
+    } catch (EntityNotFoundException e) {
+      return;
```

should this return or throw an exception?

---

return because the method is not meant to throw an error.

---

I think @Cindyzhang977 makes a good point. Throwing an exception here would be preferable, because otherwise the call to updateTrip can silently fail.

---

okidoki, method now throws EntityNotFoundException